src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs

```diff
@@ -0,0 +1,181 @@
+            // to disk.  We want to do this even when the data is unchanged so that our version
+            // stamps will be correct for the next time we come around to analyze this project.
+            //
+            // Note: we have a potential race condition here.  Specifically, for simplicity, the VS
+            // side will return immediately, without actually notifying the project system.  That
+            // means that we could persist the data to local storage that isn't in sync with what
+            // the project system knows about.  i.e. if VS is closed or crashes before that
+            // information is persisted, then these two systems will be in disagreement.  this is
+            // believed to not be a big issue given how small a time window this would be and how
+            // easy it would be to get out of that state (just edit the file).
+
+            await PersistLatestInfosAsync(projectVersion, latestInfos, cancellationToken).ConfigureAwait(false);
+        }
+
+        private async Task PersistLatestInfosAsync(VersionStamp projectVersion, (Document, DesignerInfo? info, bool changed)[] latestInfos, CancellationToken cancellationToken)
+        {
+            foreach (var (doc, info, _) in latestInfos)
+            {
+                // Skip documents that didn't change contents/version at all.  No point in writing
+                // back out the exact same data as before.
+                if (info == null)
+                    continue;
+
+                using var memoryStream = new MemoryStream();
+                using var writer = new ObjectWriter(memoryStream);
+
+                info.Value.WriteTo(writer, projectVersion);
+
+                memoryStream.Position = 0;
+                await _storage.WriteStreamAsync(
+                    doc, DataKey, memoryStream, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        private async Task<(Document, DesignerInfo? info, bool changed)[]> ComputeLatestInfosAsync(
+            Project project, VersionStamp projectVersion,
+            Document? specificDoc, CancellationToken cancellationToken)
+        {
+            var compilation = await project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
+            var designerCategoryType = compilation.DesignerCategoryAttributeType();
+
+            using var _1 = ArrayBuilder<Task<(Document, DesignerInfo?, bool changed)>>.GetInstance(out var tasks);
+            foreach (var document in project.Documents)
+            {
+                // If we're only analyzing a specific document, then skip the rest.
+                if (specificDoc != null && document != specificDoc)
+                    continue;
+
+                tasks.Add(ComputeDesignerAttributeInfoAsync(
+                    projectVersion, designerCategoryType, document, cancellationToken));
+            }
+
+            var latestInfos = await Task.WhenAll(tasks).ConfigureAwait(false);
```

Comment:
>var latestInfos = [](start = 12, length = 17)

nit: Why not just `return await Task.WhenAll(tasks).ConfigureAwait(false)`?

Reply:
it was nice for debugging :)

Reply:
The debugger should store the value of last method return into a $ReturnValue pseudo-variable that you can inspect in watch window: [https://devblogs.microsoft.com/devops/seeing-function-return-values-in-the-visual-studio-2013-debugger/](https://devblogs.microsoft.com/devops/seeing-function-return-values-in-the-visual-studio-2013-debugger/)

---
In reply to: [392394267](https://github.com/dotnet/roslyn/pull/42346#discussion_r392394267) [](ancestors = 392394267)

Reply:
removed.

Reply:
removed.